core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/hashJoin/HashJoinProbeOpExec.scala

```diff
@@ -0,0 +1,191 @@
+package edu.uci.ics.texera.workflow.operators.hashJoin
+
+import edu.uci.ics.amber.engine.architecture.worker.PauseManager
+import edu.uci.ics.amber.engine.common.InputExhausted
+import edu.uci.ics.amber.engine.common.rpc.AsyncRPCClient
+import edu.uci.ics.texera.workflow.common.operators.OperatorExecutor
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.Tuple.BuilderV2
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, Schema}
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+class HashJoinProbeOpExec[K](
+    val buildAttributeName: String,
+    val probeAttributeName: String,
+    val joinType: JoinType,
+    val buildSchema: Schema,
+    val probeSchema: Schema,
+    val outputSchema: Schema
+) extends OperatorExecutor {
+  var currentTuple: Tuple = _
+
+  var buildTableHashMap: mutable.HashMap[K, (ListBuffer[Tuple], Boolean)] = _
+
+  override def processTexeraTuple(
+      tuple: Either[Tuple, InputExhausted],
+      input: Int,
+      pauseManager: PauseManager,
+      asyncRPCClient: AsyncRPCClient
+  ): Iterator[Tuple] = {
+    tuple match {
+      case Left(tuple) =>
+        if (input == 0) {
+          buildTableHashMap.update(tuple.getField("key"), tuple.getField("value"))
+          Iterator()
+        } else {
+          // probing phase
+          val key = tuple.getField(probeAttributeName).asInstanceOf[K]
+          val (matchedTuples, _) =
+            buildTableHashMap.getOrElse(key, (new ListBuffer[Tuple](), false))
+
+          if (matchedTuples.isEmpty) {
+            // do not have a match with the probe tuple
+            if (joinType != JoinType.RIGHT_OUTER && joinType != JoinType.FULL_OUTER) {
+              return Iterator()
+            }
+            performRightAntiJoin(tuple)
+          } else {
+            // found a join match group
+            buildTableHashMap.put(key, (matchedTuples, true))
+            performJoin(tuple, matchedTuples)
```

Comment:
storing the boolean inside the buildTableHashMap may not be an ideal design. I know this is taken from the old implementation. we could refactor it later.

Reply:
This is related to Reshape. Let's do it with the next remove skew handling PR.